src/Bladeburner/Actions/BlackOperation.ts

```diff
@@ -60,4 +62,8 @@ export class BlackOperation extends ActionClass {
   getTeamSuccessBonus = operationTeamSuccessBonus;
 
   getActionTypeSkillSuccessBonus = operationSkillSuccessBonus;
+
+  getTeamCasualtiesRoll = getRandomIntInclusive;
+
+  resolveTeamCasualties = resolveTeamCasualties;
```

Comment:
This is more interface complexity that exists only for testing.

`getTeamCasualtiesRoll` obfuscates the intent a fair amount (it gets plumbed through a class and an interface to ultimately be a simple call to `getRandomIntInclusive`. However, this is balanced by the fact that it would be a bit trickier to mock around without it.

AFAICT, `resolveTeamCasualties` isn't even *used* as a test injection point, and is thus completely unneeded. If you got rid of both of these variables, you could also completely remove the `TeamActionWithCasualties` interface and `resolveTeamCasualties` could be a simple function that takes `teamCount` as a number instead of semi-awkwardly through `this`. It would also be just as straightforward to mock, if you ever needed/wanted to in the future.

EDIT: Just saw your change to fix the 0/1 issue. I guess the interface has a stay on execution. XD You *could* pass that in as another free parameter to the function, but that's getting more ugly... it tips the balance back towards it being worth it to have an interface. I still don't think you need the instance variable for the function here, though. It can just be a free function that takes a `TeamActionWithCasualties` as first parameter. (Normally you would just put it on the base class, but that doesn't work here when you don't have a base class for just this pair of classes.)

Reply:
> AFAICT, resolveTeamCasualties isn't even used as a test injection point, and is thus completely unneeded. If you got rid of both of these variables, you could also completely remove the TeamActionWithCasualties interface and resolveTeamCasualties could be a simple function that takes teamCount as a number instead of semi-awkwardly through this. It would also be just as straightforward to mock, if you ever needed/wanted to in the future.

It won't work. It needs teamCount, sleeveSize and teamSize (not to be confused with teamCount).  I tried it and this is the minimal complexity I've been able to scale it down to. It's **so easy** to confuse teamCount and teamSize and I wanted it to be handled explicitly. A narrower type (the interface) but generously wide instance (Bladeburner, Action) to make it user-friendly on the developer side who's working on CompleteAction.

I tried mocking it. But ended up with a lot more complexity (and automocking on the module level which made the tests much slower).

However, if you insist I _could_ make it a public module export on TeamCasualties and then mock that, but then the tests get exposed to much more "implementation detail" that isn't captured by the interface. It's a hard balance. Anything touching entropy needs to be controlled.

**Okay, a bit more context**
I need to be able to control the random range to test it properly. If we don't want this interface shenanigans, then I'd rather throw away the tests that are flaky when random. But that means we lose 2/3rds of the tests. The issue isn't Math.random or something similar. The problem is that our numbers cause the rolls to be exhaustive at random. It's exhaustive if casualties are 0% or 100%, and the ranges can randomly range to 50%-100% or 0%-100% depending on what the attempt roll was. This makes it incredibly difficult to test, even with monte-carlo simulations. I either need control over an non-exhaustive range **or** a forced scope to the roll (MIN,MAX). I opted for the latter, but it requires **one** injection point in production scope.

Now, I'm aware you may consider injection scope on an interface more invasive than on a package module, but from the perspective of production code, **there will be** a public function somewhere to roll the range (either exported, or on a class/interface). There is no way avoiding having it be "mentioned" in the final output. If this is a huge blocker, I'd much rather remove the tests, than try to pretend we're not affecting public scope. As I mentioned in the earlier round of reviews, this is one of those problems that cannot be **without compromise**. That's the nature of random elements.

Reply:
Edit: As a compromise if you want I can put the roll on Bladeburner instead (on the OperationTeam interface). That way it requires less care on the Action data side.

Reply:
OK, as you said it's a compromise but I like this compromise better.

However, I feel like you didn't see the other part of my comment? Specifically, regarding the `resolveTeamCasualties` function itself, that it doesn't need to be part of the interface but can be a free function that takes an interface object as a param. And that should be just as easy to mock (you can mock out the free function in the tests, if needed.)

Reply:
I misunderstood. I thought you meant the resolve as being part of the interface, not the function declaration being referenced and obsolete after the 0/1 fixed gave legit purpose for the namespace. Let me re-read.

Reply:
I did comment on it, quoted below your original. You mean to have it be a global function? Doesn't that make the invocations more verbose?

Reply:
Oh, you mean it can _stay_ on the class, but not be part of the interface? Yeah it seems we lost that structural somewhere during refactoring. I'll try it, see how it looks.

Reply:
_I still see you referring to an object with an interface but I'm not quite sure what that's highlighting. Could you give me a quick example or code pin?_

Reply:
I _think_ I got what you meant, let me know!

Reply:
Yup, that was what I was talking about!